src/ArduinoCloudProperty.hpp

```diff
@@ -65,25 +62,76 @@ enum class CborIntegerMapKey : int {
   DataValue    =  8  /* vd   */
 };
 
+template <typename T>
+class MapEntry {
+  public:
+
+    MapEntry() : _is_set(false) { }
+
+    inline void    set(T const & entry) {
+      _entry = entry;
+      _is_set = true;
+    }
+    inline T const get() const {
+      return _entry;
+    }
+
+    inline bool    isSet() const {
+      return _is_set;
+    }
+    inline void    reset()       {
+      _is_set = false;
+    }
+
+  private:
+
+    T    _entry;
+    bool _is_set;
+
+};
+
+class CborMapData {
+
+  public:
+    MapEntry<int>    base_version;
+    MapEntry<String> base_name;
+    MapEntry<double> base_time;
+    MapEntry<String> name;
+    MapEntry<String> attribute_name;
+    MapEntry<float>  val;
+    MapEntry<String> str_val;
+    MapEntry<bool>   bool_val;
+    MapEntry<double> time;
+};
+
+enum class Permission {
+  Read, Write, ReadWrite
+};
+
+enum class Type {
+  Bool, Int, Float, String
+};
+
+enum class UpdatePolicy {
+  OnChange, TimeInterval
+};
+
 typedef void(*UpdateCallbackFunc)(void);
 
 /******************************************************************************
    CLASS DECLARATION
  ******************************************************************************/
 
-template <typename T>
 class ArduinoCloudProperty {
   public:
-
-    ArduinoCloudProperty(T & property, String const & name, Permission const permission);
-
-    void writeByCloud(T const val);
+    ArduinoCloudProperty();
+    void init(String const name, Permission const permission);
 
     /* Composable configuration of the ArduinoCloudProperty class */
-    ArduinoCloudProperty<T> & onUpdate(UpdateCallbackFunc func);
-    ArduinoCloudProperty<T> & onSync(void (*func)(ArduinoCloudProperty<T> property));
-    ArduinoCloudProperty<T> & publishOnChange(T const min_delta_property, unsigned long const min_time_between_updates_millis = 0);
-    ArduinoCloudProperty<T> & publishEvery(unsigned long const seconds);
+    ArduinoCloudProperty & onUpdate(UpdateCallbackFunc func);
+    ArduinoCloudProperty & onSync(void (*func)(ArduinoCloudProperty &property));
```

Comment:
Please consider defining the sync function pointer as a `typedef` and then use the defined type instead of rewriting the whole function pointer definition on every occurence, e.g. 
```C
typedef void(*SyncCallbackFunc)(ArduinoCloudProperty &property);
/* ... */
ArduinoCloudProperty & onSync(SyncCallbackFunc func);
```

Reply:
Done